File: tree/exercise/J68.ts
// Definition for a binary tree node.
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}
 
//二叉树的左子树的所有节点都比它的根节点小，右子树的所有有节点都比它的根节点大。
//因此如果p和q比root一个小，一个大，则父节点是根节点，都小则继续往根节点的左子树变遍历，否则往根节点的右子树查找
function lowestCommonAncestor(root: TreeNode | null, p: number, q: number): TreeNode | null {
    if(!root) return null
    while(root){
        if(p < root.val && q < root.val) root = root.left
        else if(p > root.val && q > root.val)  root = root.right
        else break
    }
    return root
};

File: tree/exercise/J68.py
#!/usr/bin/python
# 根据分析所知道，p和q在么在最近节点a的两侧，则所找的就是这个节点a
# p和q在最近节点的一侧，那么找的节点就是p或q，哪个先遍历到时哪个

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if not root or root == p or root == q: return root
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        if not left: return right
        elif not right: return left
        return root

